pkg/instancetype/instancetype.go

```diff
@@ -521,6 +531,123 @@ func (m *methods) findClusterInstancetype(vm *virtv1.VirtualMachine) (*instancet
 	return instancetype, nil
 }
 
+func (m *methods) InferDefaultInstancetype(vm *virtv1.VirtualMachine) (*virtv1.InstancetypeMatcher, error) {
+	if vm.Spec.Instancetype == nil || vm.Spec.Instancetype.InferFromVolume == "" {
+		return nil, nil
+	}
+	defaultName, defaultKind, err := m.inferDefaultsFromVolumes(vm, vm.Spec.Instancetype.InferFromVolume, DefaultInstancetypeAnnotation, DefaultInstancetypeKindAnnotation)
+	if err != nil {
+		return nil, err
+	}
+	return &v1.InstancetypeMatcher{
+		Name: defaultName,
+		Kind: defaultKind,
+	}, nil
+}
+
+func (m *methods) InferDefaultPreference(vm *virtv1.VirtualMachine) (*virtv1.PreferenceMatcher, error) {
+	if vm.Spec.Preference == nil || vm.Spec.Preference.InferFromVolume == "" {
+		return nil, nil
+	}
+	defaultName, defaultKind, err := m.inferDefaultsFromVolumes(vm, vm.Spec.Preference.InferFromVolume, DefaultPreferenceAnnotation, DefaultPreferenceKindAnnotation)
+	if err != nil {
+		return nil, err
+	}
+	return &v1.PreferenceMatcher{
+		Name: defaultName,
+		Kind: defaultKind,
+	}, nil
+}
+
+func (m *methods) inferDefaultsFromVolumes(vm *virtv1.VirtualMachine, inferFromVolumeName, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	for _, volume := range vm.Spec.Template.Spec.Volumes {
+		if volume.Name != inferFromVolumeName {
+			continue
+		}
+		if volume.PersistentVolumeClaim != nil {
+			return m.inferDefaultsFromPVC(volume.PersistentVolumeClaim.ClaimName, vm.Namespace, defaultNameAnnotation, defaultKindAnnoation)
+		}
+		if volume.DataVolume != nil {
+			return m.inferDefaultsFromDataVolume(vm, volume.DataVolume.Name, defaultNameAnnotation, defaultKindAnnoation)
+		}
+		return "", "", fmt.Errorf("unable to infer defaults from Volume as type is not supported")
+	}
+	return "", "", fmt.Errorf("unable to find volume %s to infer defaults", inferFromVolumeName)
+}
+
+func inferDefaultsFromAnnotations(annotations map[string]string, defaultNameAnnotation, defaultKindAnnoation string) (string, string, error) {
+	defaultName, hasAnnotation := annotations[defaultNameAnnotation]
+	if !hasAnnotation {
+		return "", "", fmt.Errorf("unable to find required %s annotation on the volume", defaultNameAnnotation)
+	}
+	return defaultName, annotations[defaultKindAnnoation], nil
```

Comment:
An empty value in `annotations[defaultKindAnnoation]` will default to the cluster wide type?